[4.1.0] Choreo Connect graphQL documentation addition (#6127)
wso2/docs-apim: en/docs/deploy-and-publish/deploy-on-gateway/choreo-connect/deploy-api/deploy-graphql-api-in-choreo-connect.md

# Deploying a GraphQL API in Choreo Connect

GraphQL is a querying language for APIs. Initially it was developed by Facebook. When compared to the REST APIs, GraphQL APIs allow
you to query only the required data. Response will have only the data you requested (over-fetching and under-fetching will not happen). Using a single request you can fetch data
related to multiple resources. This is not possible with REST APIs. If REST API is used, you will have to consider multiple requests
to gather the same amount of data. GraphQL APIs are based on a strong type system. Schema Definition Language (SDL) file defines the
contract between client and the server. It explains how client can access the data provided by the server. This page explains how to
deploy a GraphQL API in Choero Connect and also explains the supported features of Choreo Connect for GraphQL APIs.

> !!! attention "Update Level _"

Author: need to update the version in here.

Reviewer: We should add the date of the update as well.

    This feature is available only as an update, after Update level _ and further. For more information, see [Link]({{base_path}}<>).

> Choreo Connect supports below features additionally to the basic GraphQL API invocations.

Reviewer: ```suggestion
Choreo Connect supports the below features in addition to the basic GraphQL API invocations.
```

- Runtime configurations
    - Transport level security
    - [GraphQL query analysis (query complexity and query depth limitation)]({{base_path}}/design/rate-limiting/graphql-api/overview-query-limits-for-graphql/#static-query-analyzer)
    - Cross-Origin Resource Sharing (CORS)
- [Rate limiting policies (API level and operation level) with the support of WSO2 API-M]({{base_path}}/design/create-api/create-a-graphql-api/#rate-limiting-for-graphql-operations)
- [Security scopes]({{base_path}}/design/create-api/create-a-graphql-api/#authorization-for-graphql-operations)
- Enable/ disable security for GraphQL operations

!!! note
    Currently Choreo Connect supports **`query`** and **`mutation`** GraphQL operations only.

> You can deploy GraphQL APIs using Choreo Connect by following one of the two deployment modes described in below.

Reviewer: ```suggestion
You can deploy GraphQL APIs using Choreo Connect by following one of the two deployment modes described below.
```

|**Mode**                               | **Method**    |
|---------------------------------------|---------------|
|[Choreo Connect with WSO2 API Manager as a Control Plane]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/apim-as-control-plane/)   | [Via WSO2 API Manager Publisher Portal](#via-wso2-api-manager-publisher-portal)  |
|[Choreo Connect as a Standalone Gateway]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/concepts/as-a-standalone-gateway/)                  |[Via apictl for Standalone Mode](#via-apictl-for-standalone-mode)                |

## Via WSO2 API Manager Publisher Portal

> Below instructions explain how to deploy a GraphQL to the Choreo Connect via the Publisher Portal in WSO2 API Manager.

Reviewer: ```suggestion
Below instructions explain how to deploy a GraphQL API to Choreo Connect via the Publisher Portal in WSO2 API Manager.
```

!!! info "Before you begin"

>     - This guide assumes that you already have a up and running Choreo Connect instance configured with WSO2 API Manager. If not, you can refer to the [Quick Start Guide]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/quick-start-guide-docker-with-apim/) on how to run Choreo Connect with WSO2 API Manager. To learn more about Choreo Connect related componentes, refer to the [Overview of Choreo Connect]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/choreo-connect-overview).

Reviewer: ```suggestion
    - This guide assumes that you already have an up and running Choreo Connect instance configured with WSO2 API Manager. If not, you can refer to the [Quick Start Guide]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/quick-start-guide-docker-with-apim/) on how to run Choreo Connect with WSO2 API Manager. To learn more about Choreo Connect related components, refer to the [Overview of Choreo Connect]({{base_path}}/deploy-and-publish/deploy-on-gateway/choreo-connect/getting-started/choreo-connect-overview).
```

    - GraphQL schema file (SDL) and a GraphQL server implementaion relevant to the SDL.

> ### Step 1 - Create a GraphQL API in WSO2 API Manager publisher portal

Reviewer: ```suggestion
### Step 1 - Create a GraphQL API in WSO2 API Manager Publisher Portal
```

> You can create a GraphQL API in WSO2 API-M publisher portal by following the steps described below.

Reviewer: ```suggestion
You can create a GraphQL API in WSO2 API-M Publisher Portal by following the steps described below.
```

1. Sign in to the API Publisher Portal.

    `https://<hostname>:9443/publisher`

    Example: `https://localhost:9443/publisher`

    Let's use `admin` as the username and password to sign in.

2. Click **Create API** and then click **Import GraphQL SDL**.

    [![Create GraphQL Schema Option]({{base_path}}/assets/img/learn/create-graphql-schema-option.png)]({{base_path}}/assets/img/learn/create-graphql-schema-option.png)

3. Import the schema by dragging and dropping the file or by uploading the file, and click **Next**.

>     Let's use the [StarWarsAPI schema definition](../../assets/attachments/learn/schema_graphql.graphql) to create the schema file. 

Reviewer: Use base_path for this link.

       The Deployment page appears.

       [![GraphQL API Deployment page]({{base_path}}/assets/img/learn/graphql-api-revision-1.png)]({{base_path}}/assets/img/learn/graphql-api-revision-1.png)

3. Navigate to **Publish** and click **Lifecycle**.

       The API lifecycle page appears.

4. Click **Publish** to publish the API to the API Developer Portal.

       [![Publish GraphQL API]({{base_path}}/assets/img/learn/publish-graphql-api.png)]({{base_path}}/assets/img/learn/publish-graphql-api.png)

### Step 3 - Generate an Access Token to invoke the API

1. Click **View in Dev Portal** at the top right corner to open Developer Portal in another browser tab.

2. Open **Applications** from the top menu and select **DefaultApplication** from the list.

3. Open the **Subscriptions** tab and subscribe your API to the **DefaultApplication**. 

4. Open **APIs** from the top menu and select your API.

5. Open the **Try Out** tab and click **GET TEST KEY**. This will include the access token in the cURL command you generate in the section below.

### Step 4 - Invoke the API by providing a valid GraphQL query

> 1. Since GraphQL APIs require a GraphQL query to provide results, you need to have a valid query. A sample query is provided in below section. In API Manager publisher portal, GraphiQL is used for build queries. You can use explorer option available there to build queries. More information regarding valid GraphQL query generation includes in [here]({{base_path}}/tutorials/create-and-publish-a-graphql-api/#step-51-optionally-try-out-a-query-operation).

Reviewer: ```suggestion
1. Since GraphQL APIs require a GraphQL query to provide results, you need to have a valid query. A sample query is provided in the below section. In API Manager Publisher Portal, GraphQL is used for build queries. You can use the explorer option available there to build queries. For more information regarding valid GraphQL query generation, see [here]({{base_path}}/tutorials/create-and-publish-a-graphql-api/#step-51-optionally-try-out-a-query-operation).
```

    ``` 
    query{
    human(id:1000){
        id
        name
    }
    droid(id:2000){
        name
        friends{
            name
            appearsIn
        }
    }
    }
    ```

2. Click **Execute**.

>     You will get a response for the GraphQL API invocation, as shown in below.

Reviewer: ```suggestion
    You will get a response for the GraphQL API invocation, as shown below.
```

    [![Response of GraphQL Query]({{base_path}}/assets/img/learn/graphql-response-query.png)]({{base_path}}/assets/img/learn/graphql-response-query.png)

## Via apictl for Standalone Mode

The CLI tool ([**apictl**]({{base_path}}/install-and-setup/setup/api-controller/getting-started-with-wso2-api-controller/#download-and-initialize-the-apictl)) 

> does not support initializing projects for GraphQL APIs yet. However, you can download the sample GraphQL API project in 

Reviewer: ```suggestion
does not support initializing projects for GraphQL APIs yet. However, you can download the sample GraphQL API project from 
```

> [here](https://github.com/wso2/product-microgateway/tree/main/samples/apiProjects/SampleGraphQLApi) and try out it by deploying 

Reviewer: ```suggestion
[here](https://github.com/wso2/product-microgateway/tree/main/samples/apiProjects/SampleGraphQLApi) and try it out by deploying 
```

> to the Choreo Connect. When you are using your own GraphQL API with Choreo Connect, you can change the relevant attribute values 

Reviewer: ```suggestion
it into Choreo Connect. When you are using your own GraphQL API with Choreo Connect, you can change the relevant attribute values 
```

in `api.yaml`, `graphql-complexity.yaml` and `definitions/schema.graphql` files. There is an explanation regarding those attribute values in 
[here](https://github.com/wso2/product-microgateway/tree/main/samples/apiProjects/apiProjects/SampleGraphQLApi/README.md).

> The following steps describe how to deploy GraphQL API in Choreo Connect standalone mode.

Reviewer: ```suggestion
The following steps describe how to deploy a GraphQL API in Choreo Connect standalone mode.
```

### Step 1 - Create an API Project

> When using GraphQL APIs in Choreo Connect standalone mode, you need to have an API project. Structure of a sample GraphQL API project included in below.

Reviewer: ```suggestion
When using GraphQL APIs in Choreo Connect standalone mode, you need to have an API project. The structure of a sample GraphQL API project is included below.
```

```bash
.
├── Definitions
│   ├── graphql-complexity.yaml
│   └── schema.graphql
└── api.yaml

```

!!! note
    - Similar to the sample project, you can deploy your own GraphQL API in Choreo Connect by changing the relevant attribute values that are defined in the `api.yaml` file.
    - `graphql-complexity.yaml` is an optional file where you can define complexities relevant to the GraphQL queries.

### Step 2 - Add a Choreo Connect Environment to apictl

> To use apictl with Choreo Connect, a Choreo Connect environment needs to be added to the apictl. This environment will hold the adapter URL for further commands.

Reviewer: ```suggestion
To use apictl with Choreo Connect, a Choreo Connect environment needs to be added to the apictl. This environment will hold the Adapter URL for further commands.
```

``` bash
apictl mg add env dev --adapter https://localhost:9843
```

!!! info

>     Note `mg` in the above command. The apictl commands that starts as `apictl mg` are Choreo Connect specific. If a command does not have `mg` after `apictl` then the command could probably be common to both Choreo Connect and API Manager, but it could also be API Manager specific. 

Reviewer: ```suggestion
    The apictl commands that starts as `apictl mg` are Choreo Connect specific. If a command does not have `mg` after `apictl` then the command could probably be common to both Choreo Connect and API Manager, but it could also be API Manager specific. 
```

Reviewer: Removed an unnecessary sentence here.

!!! tip

    The apictl commands here onwards are executed with the -k flag to avoid SSL verification with the Choreo Connect.

    To communicate via HTTPS without skipping SSL verification (without -k flag), add the cert in the Choreo Connect truststore into the `<USER_HOME>/.wso2apictl/certs` folder.

### Step 3 - Log in to the Choreo Connect Environment in apictl

> You can use the following command to log in to the above Choreo Connect cluster (log in to the Choreo Connect adapter). When you log in, an access token will be retrieved from Choreo Connect and it will be saved in  the apictl.

Reviewer: ```suggestion
You can use the following command to log in to the above Choreo Connect cluster (log in to the Choreo Connect Adapter). When you log in, an access token will be retrieved from Choreo Connect and it will be saved in apictl.
```

``` bash tab="Format"
apictl mg login dev -u <username> -p <password> -k
```

``` bash tab="Example"
apictl mg login dev -u admin -p admin -k
```

### Step 4 - Deploy the API

Now let's deploy the API to Choreo Connect by executing the following command.     

``` bash
apictl mg deploy api -f <path_to_the_API_project_just_created>/<gqlProjectName> -e dev -k
```  

### Step 5 - Generate an access token

After exposing the GraphQL API via Choreo Connect, you can invoke an API with a valid access token.

Let's use the following command to generate a JWT to access the API, and set it to the variable `TOKEN`. 

``` tab="Format"
TOKEN=$(curl -X POST "https://<hostname>:<port>/testkey" -d "scope=<scope values>" -H "Authorization: Basic base64(username:password)" -k -v)

```

``` tab="Example"
TOKEN=$(curl -X POST "https://localhost:9095/testkey" -d "" -H "Authorization: Basic YWRtaW46YWRtaW4=" -k -v)

```

### Step 6 - Invoke the API

!!! note

>     - Choreo Connect handles GraphQL API requests considering the `POST` HTTP method type only. Therefore, the query given in the request should follow the JSON structure explained in [here](https://graphql.org/learn/serving-over-http/#post-request).

Reviewer: ```suggestion
    - Choreo Connect handles GraphQL API requests considering the `POST` HTTP method type only. Therefore, the query given in the request should follow the JSON structure explained [here](https://graphql.org/learn/serving-over-http/#post-request).
```

>     - Request should include one of the http request `Content-Type` headers listed in below. 

Reviewer: ```suggestion
    - The request should include one of the HTTP request `Content-Type` headers listed below. 
```

        - `-H "Content-Type: application/json"`
        - `-H "Content-Type: application/graphql"`

> 1. Invoke the API using the following cURL command structure.

Reviewer: Change this to a bullet point. We use numbers if there is some sequence and in this case there is only one item in the list so a bullet is more appropriate.,
